benches/dapol.rs

```diff
@@ -1,175 +1,357 @@
+                // tree build compute time
+                group.bench_with_input(
+                    BenchmarkId::new(
+                        "build_tree",
+                        format!("{:?}/{:?}/NUM_USERS: {:?}", tup.0, tup.1, tup.2),
+                    ),
+                    &tup,
+                    |bench, tup| {
+                        bench.iter(|| {
+                            ndm_smt = Some(setup::build_ndm_smt(tup.clone()));
+                        });
+                    },
+                );
+
+                e.advance().unwrap();
+                let after = alloc.read().unwrap();
+
+                // mem used is the difference between the 2 measurements
+                let diff = after - before;
+
+                // tree build file size
+                let tree_build_file_size = setup::serialize_tree(
+                    &ndm_smt.as_ref().expect("Tree not found"),
+                    PathBuf::from("./target"),
+                );
+
+                let tree_build = Metrics {
+                    variable: Variable::TreeBuild,
+                    mem_usage: setup::bytes_as_string(diff),
+                    file_size: tree_build_file_size,
+                };
+
+                println!("\n{:?}\n", tree_build);
+            }
+        }
     }
 
-    group.finish();
+    group.finish()
 }
 
-fn generate_proof(c: &mut Criterion) {
-    let mut group = c.benchmark_group("prove");
+fn bench_generate_proof(c: &mut Criterion) {
+    let e = epoch::mib().unwrap();
+    let alloc = stats::allocated::mib().unwrap();
+
+    let mut group = c.benchmark_group("dapol");
     group.sample_size(10);
+    // `SamplingMode::Flat` is used here as that is what Criterion recommends for long-running benches
+    // https://bheisler.github.io/criterion.rs/book/user_guide/advanced_configuration.html#sampling-mode
+    group.sampling_mode(SamplingMode::Flat);
+
+    dapol::initialize_machine_parallelism();
+
+    let thread_counts = {
+        let mut tc: Vec<u8> = Vec::new();
+
+        let max_thread_count: u8 = MaxThreadCount::default().get_value();
+
+        let step = if max_thread_count < 8 {
+            1
+        } else {
+            max_thread_count >> 2
+        };
+
+        for i in (step..max_thread_count).step_by(step as usize) {
+            tc.push(i);
+        }
+
+        tc.push(max_thread_count);
+
+        tc
+    };
+
+    let mut rng = rand::thread_rng();
+
+    e.advance().unwrap();
+
+    for h in TREE_HEIGHTS.into_iter() {
+        for t in thread_counts.iter() {
+            for u in NUM_USERS.into_iter() {
+                let mut proof = Option::<InclusionProof>::None;
+
+                // Many of the statistics tracked by `jemalloc` are cached.
+                // The epoch controls when they are refreshed.
+                // We care about measuring proof generation so we refresh before it's construction
+                e.advance().unwrap();
+                let before = alloc.read().unwrap();
```

Comment:
No need to do memory readings for proof generation because it won't take up very much memory.